L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,415 @@
+    const double H12 = H12_->digi(state->H12());
+    // cov. matrix
+    double C00 = C00_->digi(state->C00());
+    double C01 = C01_->digi(state->C01());
+    double C11 = C11_->digi(state->C11());
+    double C22 = C22_->digi(state->C22());
+    double C23 = C23_->digi(state->C23());
+    double C33 = C33_->digi(state->C33());
+    // stub phi residual wrt current state
+    const double r0C = x1_->digi(m0 - x1);
+    const double r0 = r0_->digi(r0C - x0 * H00);
+    // stub z residual wrt current state
+    const double r1C = x3_->digi(m1  - x3);
+    const double r1 = r1_->digi(r1C - x2 * H12);
+    // matrix S = H*C
+    const double S00 = S00_->digi(C01  + H00 * C00);
+    const double S01 = S01_->digi(C11  + H00 * C01);
+    const double S12 = S12_->digi(C23 + H12 * C22);
+    const double S13 = S13_->digi(C33 + H12 * C23);
+    // Cov. matrix of predicted residuals R = V+HCHt = C+H*St
+    const double R00C = S01_->digi(v0 + S01);
+    const double R00 = R00_->digi(R00C + H00 * S00);
+    const double R11C = S13_->digi(v1 + S13);
+    const double R11 = R11_->digi(R11C + H12 * S12);
+    // imrpoved dynamic cancelling
+    const int msb0 = max(0, (int)ceil(log2(R00 / R00_->base())));
+    const int msb1 = max(0, (int)ceil(log2(R11 / R11_->base())));
+    const double R00Rough = R00Rough_->digi(R00 * pow(2., 16 - msb0));
+    const double invR00Approx = invR00Approx_->digi(1. / R00Rough);
+    const double invR00Cor = invR00Cor_->digi(2. - invR00Approx * R00Rough);
+    const double invR00 = invR00_->digi(invR00Approx * invR00Cor * pow(2., 16 - msb0));
+    const double R11Rough = R11Rough_->digi(R11 * pow(2., 16 - msb1));
+    const double invR11Approx = invR11Approx_->digi(1. / R11Rough);
+    const double invR11Cor = invR11Cor_->digi(2. - invR11Approx * R11Rough);
+    const double invR11 = invR11_->digi(invR11Approx * invR11Cor * pow(2., 16 - msb1));
+    // Kalman gain matrix K = S*R(inv)
+    const double K00 = K00_->digi(S00 * invR00);
+    const double K10 = K10_->digi(S01 * invR00);
+    const double K21 = K21_->digi(S12 * invR11);
+    const double K31 = K31_->digi(S13 * invR11);
+    // Updated helix params & their cov. matrix
+    x0 = x0_->digi(x0 + r0 * K00);
+    x1 = x1_->digi(x1 + r0 * K10);
+    x2 = x2_->digi(x2 + r1 * K21);
+    x3 = x3_->digi(x3 + r1 * K31);
+    C00 = C00_->digi(C00 - S00 * K00);
+    C01 = C01_->digi(C01 - S01 * K00);
+    C11 = C11_->digi(C11 - S01 * K10);
+    C22 = C22_->digi(C22 - S12 * K21);
+    C23 = C23_->digi(C23 - S13 * K21);
+    C33 = C33_->digi(C33 - S13 * K31);
+    // update variable ranges to tune variable granularity
+    C22_->updateRangeActual(C22);
+    C23_->updateRangeActual(C23);
+    C33_->updateRangeActual(C33);
+    C00_->updateRangeActual(C00);
+    C01_->updateRangeActual(C01);
+    C11_->updateRangeActual(C11);
+    // create updated state
+    states_.emplace_back(State(state, (initializer_list<double>){x0, x1, x2, x3, C00, C11, C22, C33, C01, C23}));
```

Comment:
As the update of states_ & state are a key part of the KF update, whereas the calls to updateRangeActual() are just used for digi tuning, please move updates of states_ & state in front of all calls to updateRangeActual().